src/forcing/ForcingsEngine.cpp

```diff
@@ -0,0 +1,282 @@
+#include "ForcingsEngine.hpp"
+
+namespace data_access {
+
+std::unordered_map<std::string, std::unique_ptr<ForcingsEngine>> ForcingsEngine::instances_{};
+
+ForcingsEngine::ForcingsEngine(const std::string& init, size_type time_start, size_type time_end)
+  : time_start_(std::chrono::seconds{time_start})
+  , time_end_(std::chrono::seconds{time_end})
+{
+    check_runtime_requirements();
+
+    bmi_ = std::make_unique<models::bmi::Bmi_Py_Adapter>(
+        "ForcingsEngine",
+        init,
+        "NextGen_Forcings_Engine.BMIForcingsEngine",
+        true, /* allow_exceed_end */
+        true, /* has_fixed_time_step */
+        utils::getStdOut()
+    );
+    
+    time_step_ = std::chrono::seconds{static_cast<int64_t>(bmi_->GetTimeStep())};
+    var_outputs_ = bmi_->GetOutputVarNames();
+
+    auto cat_var_pos = std::find(var_outputs_.begin(), var_outputs_.end(), "CAT-ID");
+    if (cat_var_pos == var_outputs_.end()) {
+        throw std::runtime_error{
+            "Failed to initialize ForcingsEngine: `CAT-ID` is not an output variable of the forcings engine."
+            " Is it running with `GRID_TYPE` set to 'hydrofabric'?"
+        };
+    }
+    var_outputs_.erase(cat_var_pos); // Erase "CAT-ID" from outputs
+
+
+    const auto time_dim = static_cast<size_type>((time_end_ - time_start_) / time_step_);
+    const auto id_dim   = static_cast<size_type>(bmi_->GetVarNbytes("CAT-ID") / bmi_->GetVarItemsize("CAT-ID"));
+    const auto var_dim  = var_outputs_.size();
+
+    bmi_->Update();
+    time_current_index_++;
+    
+    const auto* ptr = static_cast<int*>(bmi_->GetValuePtr("CAT-ID"));
+    var_divides_ = std::vector<int>(ptr, ptr + id_dim);
+    var_cache_ = decltype(var_cache_){{ time_dim, id_dim, var_dim }};
+
+    // Cache initial iteration
+    const auto size = var_divides_.size();
+    for (size_type vi = 0; vi < var_outputs_.size(); vi++) {
+        // Get current values for each variable
+        const auto var_span = boost::span<const double>{
+            static_cast<double*>(bmi_->GetValuePtr(var_outputs_[vi])),
+            size
+        };
+
+        for (size_type di = 0; di < size; di++) {
+            // Set current variable value for each divide
+            var_cache_.at({{time_current_index_, di, vi}}) = var_span[di];
+        }
+    }
+
+    time_current_index_++;
+}
+
+ForcingsEngine::~ForcingsEngine()
+{
+    finalize();
+}
+
+ForcingsEngine& ForcingsEngine::instance(
+    const std::string& init,
+    std::size_t time_start,
+    std::size_t time_end
+)
+{
+    if (instances_[init] == nullptr) {
+        instances_[init] = std::make_unique<ForcingsEngine>(init, time_start, time_end);
+    }
```

Comment:
Is it valid to ask for a `ForcingsEngine` instance with an `init` matching some other request, but different time arguments? Given the discussion about checking begin/end time, and possibly extending data, that seems to invite trouble.

Reply:
The apparent situation that comes to mind is if we are running a separate workflow under a new temporal domain in the same execution space. To me, this seems like it should be an invalid call.

I can add two asserts to ensure the time arguments match that of the instance received, so we can see if this situation (or similar) occurs and address it then.

Reply:
The asserts of matching time arguments sound like a good answer for now.